Jayemar/hashtable for files cache

jayemar opened the pull request:
This PR introduces a hashmap to track the links, tags, and aliases associated with each file.  This allows for a more responsive Obsidian workflow when viewing and editing note metadata and jumping between notes.

jayemar commented on obsidian.el:
```diff
@@ -6,8 +6,7 @@
 ;; Version: 1.4.4
-;; Package-Requires: ((emacs "27.2") (f "0.2.0") (s "1.12.0") (dash "2.13") (markdown-mode "2.5") (elgrep "1.0.0") (yaml "0.5.1"))
-
+;; Package-Requires: ((emacs "27.2") (f "0.2.0") (s "1.12.0") (dash "2.13") (markdown-mode "2.5") (elgrep "1.0.0") (yaml "0.5.1") (ht "2.3"))
```
Require the `ht` library for hashtable functions

jayemar commented on obsidian.el:
```diff
@@ -94,6 +94,23 @@
   "Daily notes' template filename in templates directory"
   :type 'file)
 
+(defcustom obsidian--use-update-timer t
```
Creates a timer to periodically check for file updates

jayemar commented on obsidian.el:
```diff
@@ -79,21 +79,37 @@
+  "Daily notes' template filename in templates directory."
   :type 'file)
 
+(defcustom obsidian-use-update-timer t
```
Set a timer to periodically update the vault cache

jayemar commented on obsidian.el:
```diff
@@ -112,67 +128,165 @@ of `dirctory-files'."
-  "Clears aliases map."
-  (interactive)
-  (setq obsidian--aliases-map (make-hash-table :test 'equal)))
+(defvar obsidian--vault-cache nil
```
This is the new data structure for tracking obsidian files and their associated links, tags, and aliases

jayemar commented on obsidian.el:
```diff
@@ -284,86 +382,81 @@ Return nil if the front matter does not exist, or incorrectly delineated by
-  (->> (obsidian-list-all-tags)
-       (setq obsidian--tags-list))
-  (message "Obsidian tags updated"))
+(defun obsidian-tags-ht ()
```
Create a tags hashmap using the existing vault cache

jayemar commented on obsidian.el:
```diff
@@ -284,86 +382,81 @@ Return nil if the front matter does not exist, or incorrectly delineated by
+               obsidian--tags-map)
+      obsidian--tags-map)))
+
+(defun obsidian-tags ()
```
Use the vault cache to return a list of all tags

jayemar commented on obsidian.el:
```diff
@@ -284,86 +382,81 @@ Return nil if the front matter does not exist, or incorrectly delineated by
+                        (gethash 'tags val-map))
+                      (hash-table-values obsidian--vault-cache))))))
+
+(defun obsidian--buffer-metadata ()
```
Grab all metadata from a file and use it to update vault cache

jayemar commented on obsidian.el:
```diff
@@ -373,36 +466,53 @@ lower and upper case versions of the tags."
+  (let ((all-files (directory-files-recursively obsidian-directory "\.*$")))
+    (-filter #'obsidian-file-p all-files)))
+
+(defun obsidian-clear-cache ()
```
`obsidian-update` is now a much more lightweight function; `obsidian-clear-cache` can be used to force a cache purge

jayemar commented on obsidian.el:
```diff
@@ -373,36 +466,53 @@ lower and upper case versions of the tags."
+  (interactive)
+  (setq obsidian--vault-cache nil))
+
+(defun obsidian-populate-cache ()
```
`obsidian-update` is now a much more lightweight function, but `obsidian-populate-cache` can be used to repopulate all data for all files

jayemar commented on obsidian.el:
```diff
@@ -636,8 +791,22 @@ Opens markdown links in other window if ARG is non-nil.."
+          obsidian--prepare-file-path
+          (obsidian-find-point-in-file 0 arg)))))
+
+(defun obsidian-follow-backlink-at-point ()
```
Because we're now including not only the link itself but also the location of the link in the target file, we can jump directly to that location in the target file.

jayemar commented on obsidian.el:
```diff
@@ -717,17 +882,41 @@ See `markdown-follow-link-at-point' and
     (message "Template variables replaced and inserted to the buffer")
     (goto-char m)))
 
+(defun obsidian--backlinks-completion-fn (hmap)
```
Show not only the target file but also the link text

licht1stein commented:
This looks like a new major version :)

licht1stein commented:
This is amazing work, thank you! Do you think there's a risk of breaking people's workflows enough to call this a 2.0.0?

jayemar commented:
Ha well my intention was that these changes wouldn't necessitate a new version, but looking through the diff again I see that I've removed and renamed some functions that some people are probably using, so I think it probably does.

Some of the function renaming I just did while I was doing some refactoring to better understand the code.  I thought I'd changed those back, but I definitely see plenty that I missed.  I could change those back to the previous names if you'd like, but the functionality for some of the removed functions no longer exists.

**Renamed functions:**
`obsidian-list-all-files`         -> `obsidian-files`
`obsidian-list-all-directories`   -> `obsidian-directories`
`obsidian-list-all-tags`          -> `obsidian-tags`
`obsidian-find-yaml-front-matter` -> `obsidian--find-yaml-front-matter-in-string`
`obsidian-prepare-file-path`      -> `obsidian--prepare-file-path`

**Renamed to private:**
`obsidian-dot-file-p`         -> `obsidian--dot-file-p`
`obsidian-user-directory-p`   -> `obsidian--user-directory-p`
`obsidian-not-trash-p`        -> `obsidian--not-trash-p`
`obsidian-not-dot-obsidian-p` -> `obsidian--not-dot-obsidian-p`

**Removed variables:**
`obsidian-files-cache`
`obsidian-cache-timestamp`
`obsidian-cache-expiry`

**Removed functions:**
`obsidian-descendent-of-p`
`obsidian-cache-needs-reset-p`
`obsidian-reset-cache`
`obsidian-read-file-or-buffer`
`obsidian-find-tags`
`obsidian-get-yaml-front-matter`
`obsidian--file-front-matter`
`obsidian--update-from-front-matter`
`obsidian--update-all-from-front-matter`
`obsidian-find-tags-in-file`
`obsidian-update-tags-list`
`obsidian-tags-backend`
`obsidian--elgrep-get-context`
`obsidian--mention-link-to-p`
`obsidian--find-links-to-file`
`obsidian--completing-read-for-matches`

**Changed function signature:**
`obsidian-backlink-jump`: `()` -> `(&optional file)`

**Changed functionality:**
`obsidian-update`: checks for new or deleted files; no longer rebuilds the entire cache

Let me know what you think. And let me know if you have thoughts on any of the naming.  I'm still a little unclear on when to use the double hyphen "private" naming convention, and if we're going to make this a version 2.0 I might as well try to get that right.

licht1stein commented:
I think we should make this 2.0, leave the names the way you see fit and add backlinks panel you mentioned :)

jayemar commented:
Okay, great!  I'll get the code for the backlinks panel cleaned up and pull that in.

I might also pull in some of the open PRs I'd created for additional testing; I created those when I was doing this work to try for more checks that I wasn't breaking things.

I've also got a branch that updates the tags regex to better align with Obsidian rules for tags, and also better handles cases where the YAML parser throws exceptions when attempting to parse front matter.  It uses some of these updated functions but should be an easy merge for the 2.0 version, so I'll get a PR created for that, too.

jayemar commented on tests/test-obsidian.el:
```diff
@@ -2,35 +2,43 @@
+;; Using a relative path for obsidian-change-vault/obsidian-change-vault will
+;; result in different values for obsidian-directory depending on the directory
+;; of the most recently visited file
+(defvar obsidian--test-dir (expand-file-name "./tests/test_vault"))
```
Using a relative path here was causing issues when running tests after other tests that visited files because the relative location would then be the location of the opened note and not the vault root.

jayemar commented on tests/test-obsidian.el:
```diff
@@ -2,35 +2,43 @@
-  (before-all (obsidian-specify-path obsidian--test-dir))
-  (after-all (obsidian-specify-path obsidian--test--original-dir))
-
+  (before-all (obsidian-change-vault obsidian--test-dir))
```
`obsidian-change-vault` updates the cache as well as changing the vault path, so `obsidian-specify-path` was replaced with that function.

jayemar commented on obsidian.el:
```diff
@@ -69,110 +68,292 @@
-When run interactively asks user to specify the path."
-  (interactive)
-  (->> (or path (read-directory-name "Specify path to Obsidian folder"))
+You most likely want to run `obsidian-change-vault'."
```
This function changes the vault path but doesn't update the cache, so the function `obsidian-change-vault` will probably be more useful for most applications.

jayemar commented on obsidian.el:
```diff
@@ -69,110 +68,292 @@
+
+(defvar obsidian--debug-messages nil "Additional messages will be displayed for debugging.")
+
+(defvar obsidian--tag-regex
```
The tag regex was updated to better align with what Obsidian Notes recognizes as valid tags.

jayemar commented on obsidian.el:
```diff
@@ -474,6 +702,68 @@ replaced by the link."
+                   (company-begin-backend 'obsidian-tags-backend)
+                 (error "Company not installed")))
+  (cl-case command
+    (prefix (and (eq major-mode 'markdown-mode)
```
As `obsidian--tag-regex` was updated, the tags backend needed to be updated, as well.  I think this still works as it did previously.

jayemar commented:
Alright, I think this branch is ready.  I've pulled in the backlinks dashboard and made some other changes based on the new cache data structure.

I wrote some additional tests to try to verify that I wasn't breaking things, and also went through the README to try to make sure I wasn't missing anything, and made some updates to that doc in the process.  It's a lot of changes, so if there are things you don't like or that you'd like to see done differently let me know.

jayemar commented:
Not quite. I recently found a bug that can prevent wiki links from being recognized so I need to get that fixed. I'll let you know when I get that fix in at which point the branch should be ready to merge.

licht1stein commented:
Thank you!

jayemar commented:
Alright @licht1stein this should now be ready to merge.

The initially loading of the cache is a bit slower than I'd like, but once it's loaded everything seems quite responsive.  And there is messaging to the user letting them know that the cache is being populated.

licht1stein commented:
We need a few people to test it out, how could we achieve this?

rross101 commented:
I'm following this thread and happy to test it; if I put this in my init.el will it pull the right commit?
```
 (package! obsidian
   :recipe (:host github
            :repo "jayemar/obsidian.el"
            :branch "jayemar/hashtable-for-files-cache"))
```

jayemar commented:
@rross101 this info looks correct but I'm not familiar with `package!`.  Which package provide that? It looks very useful.

rross101 commented:
> @rross101 this info looks correct but I'm not familiar with `package!`. Which package provide that? It looks very useful.

I'm on Doom Emacs (the only Emacs I know!) and it's the package system built in... But I've not done more complex package management like this.

jayemar commented:
> I'm on Doom Emacs (the only Emacs I know!) and it's the package system built in... But I've not done more complex package management like this.

Ah okay, I haven't run Doom in a few years, but assuming that config installs the `jayemar/hashtable-for-files-cache` from my fork of the repo then you should have the right commit.

And thanks for offering to help test!  I'm very curious to get feedback.

rross101 commented:
So, I've quite a big vault - 3460 files - it starts indexing but gets stuck at 8% and emacs freezes. Is there a way I can get more debug info?

jayemar commented:
> So, I've quite a big vault - 3460 files - it starts indexing but gets stuck at 8% and emacs freezes. Is there a way I can get more debug info?

I would add the following as the first line of the `obsidian-add-file` function so that you can get an idea of which file it might be getting stuck on:
```emacs-lisp
(f-append-text (format "Adding file %s\n" file) 'utf-8 "/tmp/debug.txt")
```

Your vault is about 5x bigger than my vault, so I just copied my vault 10x into a new directory as used that as my new vault.  It did take about a minute to populate everything, but it didn't eventually load with a problem, so I'm curious to see what could be the issue with yours.

rross101 commented:
Ah interesting it was one particular file; it wasn't that important, deleted it and it's working smoothly now.

jayemar commented:
Any chance you could send me the file? I'd be curious to know exactly what was causing the issue so that maybe we could write something to handle it better.

rross101 commented:
> Any chance you could send me the file? I'd be curious to know exactly what was causing the issue so that maybe we could write something to handle it better.

Yes, [here](https://gist.github.com/rross101/d7705c17785b599473bd2bee4a241826)

jayemar commented:
This was very helpful, thanks!  It turns out there's a bug where we enter an infinite loop when parsing a file for links if the file 1) ends with a link and 2) does not end with a final newline.  The file you sent fits both of these criteria so parsing the file was sending Emacs into an infinite loop.  If you open that file, add a newline to the end and save it, you can add it back to your vault and everything should work fine.

I thought I'd seen something similar before, and even changed some of the test files to end with links.  But it seems that many text editors force a final newline, probably because many programming languages require a final newline in their source files.  Emacs has a setting `require-final-newline` that I can't manage to keep set to nil because there are other things setting it back to t, and so saving files in Emacs always adds the final newline which prevents the infinite loop.

But now that I finally understand the conditions that cause this, I'll see if I can figure out how to prevent it.

rross101 commented:
Thanks for explaining. I'm glad to have helped.

jayemar commented:
I Just pushed a bug fix for the bug that was causing Emacs to hang for rross101 when loading their vault

rhburrows commented:
I found this pull request when I ran into #99 because I have files with empty tags in the front matter. Just wanted to let y'all know I've switched to this branch down for local use and everything seems to be working great for me so far (though I don't have a super involved workflow). Let me know if there's anything extra I can do to help test further

licht1stein commented:
@jayemar are you ready to merge? :)

jayemar commented:
@licht1stein your previous message asked me if I was ready to merge, but just to be clear I can't actually merge anything because someone else needs to review and approve the PR. Not trying to rush anything, just making sure we're on the same page.

licht1stein commented:
@jayemar it's yours to merge :) sorry for delay
